AppCenter/AppCenter/Internals/HttpClient/MSHttpClient.m

```diff
@@ -2,20 +2,50 @@
 // Licensed under the MIT License.
 
 #import "MSHttpClient.h"
+#import "MSAppCenterErrors.h"
+#import "MSAppCenterInternal.h"
+#import "MSConstants+Internal.h"
+#import "MSHttpCall.h"
+#import "MSHttpClientPrivate.h"
+#import "MSHttpUtil.h"
+#import "MSLoggerInternal.h"
+#import "MSUtility+StringFormatting.h"
+#import "MS_Reachability.h"
 
 @implementation MSHttpClient
 
-- (instancetype)init {
+#define DEFAULT_RETRY_INTERVALS @[ @10, @(5 * 60), @(20 * 60) ]
 
-  // TODO implement this properly.
-  return self = [super init];
+- (instancetype)init {
+  return [self initWithMaxHttpConnectionsPerHost:nil
+                                  retryIntervals:DEFAULT_RETRY_INTERVALS
+                                    reachability:[MS_Reachability reachabilityForInternetConnection]];
 }
 
 - (instancetype)initWithMaxHttpConnectionsPerHost:(int)maxHttpConnectionsPerHost {
+  return [self initWithMaxHttpConnectionsPerHost:@(maxHttpConnectionsPerHost)
+                                  retryIntervals:DEFAULT_RETRY_INTERVALS
+                                    reachability:[MS_Reachability reachabilityForInternetConnection]];
+}
+
+- (instancetype)initWithMaxHttpConnectionsPerHost:(NSNumber *)maxHttpConnectionsPerHost
+                                   retryIntervals:(NSArray *)retryIntervals
+                                     reachability:(MS_Reachability *)reachability {
   if ((self = [super init])) {
+    _sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+    if (maxHttpConnectionsPerHost) {
+      _sessionConfiguration.HTTPMaximumConnectionsPerHost = [maxHttpConnectionsPerHost intValue];
+    }
+    _session = [NSURLSession sessionWithConfiguration:_sessionConfiguration];
+    _pendingCalls = [NSMutableSet new];
+    _retryIntervals = [NSArray arrayWithArray:retryIntervals];
```

Comment:
Can we make this property a `copy` instead of using `arrayWithArray` or that wasn't the only reason?

Reply:
But does making it `copy` copy the original parameter? I thought `copy` means it will be copied if it's on the right-hand side of the assignment, but nothing if it's on the left.

Reply:
i think i had it backwards. will fix

Reply:
Wait actually forget about this comment that only works when using the accessor for the property so outside of the init.
But yeah copy would call NSCopying on the right object to assign the copy to the left object.